docs/nodemcu-lrm.md

```diff
@@ -0,0 +1,335 @@
+# NodeMCU Lua Reference
+
+## Introduction
+
+NodeMCU firmware is an IoT project ("the Project") which implements a Lua-based runtime for SoC modules based on the Espressif ESP8266 and ESP32 architectures.  This reference specifically addresses how the NodeMCU Lua implementation relates to standard Lua as described in the two versions of the Lua language that we currently support:
+
+-  The [Lua 5.1 Reference Manual]( https://www.lua.org/manual/5.1/) and
+-  The [Lua 5.3 Reference Manual]( https://www.lua.org/manual/5.3/) (**LRM**)
+
+Developers using the NodeMCU environment should familiarise themselves with the 5.3 LRM.
+
+The Project provides a wide range of standard library modules written in both C and Lua to support many ESP hardware modules and chips, and these are documented in separation sections in our [online documentation](https://nodemcu.readthedocs.io/).
+
+This reference supplements LRM content and module documentation by focusing on the differences between NodeMCU Lua and standard Lua 5.3 in use.
+
+One of our goals in introducing Lua 5.3 support was to maintain the continuity of our existing C modules by ensuring that they can be successfully compiled and executed in both the Lua 5.1 and 5.3 environments.  This goal was achieved by combination of:
+-  enabling relevant compatibility options for standard Lua libraries;
+-  regressing some Lua 5.3 API enhancements back into our Lua 5.1 implementation, and 
+-  making some small changes to the module source to ensure that incompatible API use is avoided.
+
+Further details are given in the [Lua compatibility](#lua-compatibility) section below.  Notwithstanding this, the Project has now deprecated Lua 5.1 and will soon be moving this version into frozen support.
```

Comment:
> moving this version into frozen support

Do we have consensus on that yet? I for one would be happy to remove once it's been deprecated for 2-3 release cycles.

Reply:
> Do we have consensus on that yet?

I feel that the committers actively involved in this process (you, me, Nathaniel, Gregor and perhaps Philip) would like to see this sooner rather than later.  Like you, I feel that it will probably take a couple of release cycles to sort out residual gremlins and round out both the documentation and NodeMCU extras.  One side effect of going for minimal change to the Lua53 core + aligning the 5.1 and 5.3 APIs over this last year is that a lot of the nasty bits of the `eLua` implementation have been removed without loss of functionality.  

No doubt they could be with some further polishing, but with these two new reference documents we at least have a formal specification of NodeMCU Lua is and advice to new C module developers can now sensibly boil down to (i) read and understand the reference manuals and (ii) start with the Lua and modules libraries for examples.

Somewhere along this process we are going to have to grasp the nettle of merging all of these enhancement into (and integrating) the ESP32 codebase, but to be honest I would be a lot more comfortable starting this if I had committed support from one of the more ESP32-aware committers.

Reply:
> I'm trying to schedule some work time to spend on the next step(s) on the 8266 & 32 branch merging work.

Fantastic.  I suspect that this effort is going to be resourced by the usual suspects: You, me, Nathaniel and Gregor, with Marcel doing his usual sanity check oversight.  

There are some fundamental architectural enhancements that I want to add to Lua 5.3 (such as on ESP building of LFS images, and dual LFS images) that I feel that we should deploy on the ESP8266 platform soon and get stable before we do the actual branch merge.  (See this [\#3128 comment](3128#issuecomment-638452596) and following for discussion about the advantages of a dual LFS.)

Even so (so long I can continue to dodgy this Covid bullet),  we should plan to do this within the next few months.